src/main/java/com/google/sps/servlets/AuthenticateServlet.java

```diff
@@ -44,9 +44,31 @@ public void doGet(HttpServletRequest request, HttpServletResponse  response) thr
     response.setContentType("application/json;");
     Gson gson = new Gson();
 
-    GivrUser user = GivrUser.getLoggedInUser();
+    GivrUser user = GivrUser.getCurrentLoggedInUser();
+
+    boolean doesUserIdExistInDatastore = GivrUser.checkIfUserWithPropertyExists("userId", user.getUserId());
+    if (!doesUserIdExistInDatastore) {
+
+      boolean doesUserEmailExistInDatastore = GivrUser.checkIfUserWithPropertyExists("userEmail", user.getUserEmail());
+      if (doesUserEmailExistInDatastore) {
+        // User is either a Moderator or Maintainer, that has not logged in.
+
+        boolean isMaintainer = user.isMaintainer();
+        if (!isMaintainer) { // User is a Moderator.
+
+          // TODO: Update organization.
+        }
+
+        // Update entity with current user's userId.
+        GivrUser.updateUserInDatastore("userEmail", user.getUserEmail(), "userId", user.getUserId());
+      }
+    } else {
+      // Ensures that the userEmail is up-to-date.
```

Comment:
should you check if the userEmail != the stored email? Otherwise you'll have extra writes and writes are more expensive than reads.